### PR TITLE
T-13259 feat: added latency collection to collector

### PR DESCRIPTION
**Pull request by commoddity**
Adds collection of the previous day's latency results to the Relay Meter, stored in a Postgres table with each row storing a single hour's daily latency (to five decimal places of a second, ie. 2 hundredths of a millisecond).

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,3 +1,6 @@
+test:
```
Comment (adshmh) [inline Makefile:1]: If we are adding make targets, I recommend adding a default target as well.

File: collector/collector.go [hunk 2]
```diff
@@ -78,11 +80,22 @@ func (c *collector) Collect(from, to time.Time) error {
 }
 
 func (c *collector) CollectTodaysMetrics() error {
+	todaysLatencies, err := c.Source.TodaysLatencies()
+	if err != nil {
+		return err
```
Comment (adshmh) [inline collector/collector.go]: Perhaps an error in latency data retrieval should not cause the collection of all other metrics to be skipped.

File: collector/collector.go [hunk 3]
```diff
@@ -78,11 +80,22 @@ func (c *collector) Collect(from, to time.Time) error {
...
+	}
+	c.Logger.WithFields(logger.Fields{"todays_latencies_count": len(todaysLatencies)}).Info("Collected todays latencies")
+
 	todaysCounts, err := c.Source.TodaysCounts()
 	if err != nil {
 		return err
 	}
-	c.Logger.WithFields(logger.Fields{"todays_metrics_count": len(todaysCounts)}).Info("Collected todays metrics")
+	c.Logger.WithFields(logger.Fields{"todays_usage_count": len(todaysCounts)}).Info("Collected todays usage")
```
Comment (adshmh) [inline collector/collector.go]: Thank you for adding a log entry for this!

File: collector/collector.go [hunk 4]
```diff
@@ -78,11 +80,22 @@ func (c *collector) Collect(from, to time.Time) error {
...
+
+	err = c.Writer.WriteTodaysLatencies(todaysLatencies)
+	if err != nil {
```
Comment (adshmh) [inline collector/collector.go:101]: Same as previous comment re errors encountered processing latency data: I don't think other metrics should be skipped. 

In addition, this function returns multiple, not necessarily related, errors: it likely needs a refactor: e.g. `WriteTodaysMetrics` as a single function?

File: db/influx.go [hunk 5]
```diff
@@ -172,6 +175,68 @@ func (i *influxDB) TodaysCounts() (map[string]api.RelayCounts, error) {
...
 }
 
+func (i *influxDB) TodaysLatencies() (map[string][]api.Latency, error) {
+	client := influxdb2.NewClient(i.Options.URL, i.Options.Token)
+	queryAPI := client.QueryAPI(i.Options.Org)
+
+	latencies := make(map[string][]api.Latency)
+	// TODO: send queries in parallel
+	query := fmt.Sprintf("from(bucket: %q)", i.Options.CurrentBucket) +
+		fmt.Sprintf(" |> range(start: %s)", startOfDay(time.Now()).Format(time.RFC3339)) +
```
Comment (adshmh) [inline db/influx.go]: I know this issue was already there in the code, but there must be a more readable way to deal with the query and its parameters. One potential solution is use of `text/template` package to make the query readable while allowing variable substitution.

File: db/postgres.go [hunk 6]
```diff
@@ -221,6 +224,51 @@ func (p *pgClient) WriteTodaysUsage(counts map[string]api.RelayCounts) error {
...
+	_, deleteErr := tx.ExecContext(ctx, "DELETE FROM todays_app_latencies")
+	if deleteErr != nil {
+		if rollbackErr := tx.Rollback(); rollbackErr != nil {
+			fmt.Printf("delete failed: %v, unable to rollback: %v\n", deleteErr, rollbackErr)
+			return deleteErr
+		}
+	}
+
+	// TODO: bulk insert
+	for app, latencies := range latencies {
```
Comment (adshmh) [inline db/postgres.go]: This variable naming works, but is confusing due to using the same name twice.

File: db/postgres.go [hunk 7]
```diff
@@ -221,6 +224,51 @@ func (p *pgClient) WriteTodaysUsage(counts map[string]api.RelayCounts) error {
...
+
+		latenciesString, err := json.Marshal(latencies)
```
Comment (adshmh) [inline db/postgres.go]: We are writing this to a postgres DB. This is trying to store multiple value in a single column, which I think should be avoided. Any other solutions, e.g. multiple rows per application (preferable) or a set number of columns to store a limited number of relays per application, works for me (as long as it conforms to relational data model)

**Review by adshmh — CHANGES_REQUESTED** [review]
Thank you for the PR. Only a few comments to address.

File: db/postgres.go (hunk 6 above)
Comment (commoddity) [reply db/postgres.go]: Fixed

File: db/influx.go (hunk 5 above)
Comment (commoddity) [reply db/influx.go]: What would you think of this approach: 
image

File: collector/collector.go (hunk 4 above)
Comment (commoddity) [reply collector/collector.go:101]: Updated

File: collector/collector.go (hunk 2 above)
Comment (commoddity) [reply collector/collector.go]: Updated to prevent failure or either counts or latency collection from stopping collection of other metrics

File: db/postgres.go (hunk 7 above)
Comment (commoddity) [reply db/postgres.go]: Updated to store each latency result to a single row in the table

File: api/meter.go [hunk 8]
```diff
@@ -35,13 +35,20 @@ type RelayMeter interface {
 	// LoadBalancerRelays returns the metrics for an Endpoint, AKA loadbalancer
 	LoadBalancerRelays(endpoint string, from, to time.Time) (LoadBalancerRelaysResponse, error)
 	AllLoadBalancersRelays(from, to time.Time) ([]LoadBalancerRelaysResponse, error)
+	AppLatency(app string) (AppLatencyResponse, error)
+	AllAppsLatency() ([]AppLatencyResponse, error)
```
Comment (adshmh) [inline api/meter.go]: Perhaps `AllLatencies` or `AllAppsLatencies` or `Latencies` are better candidates for the name of this method?

File: api/meter.go [hunk 9]
```diff
@@ -151,6 +168,13 @@ func (r *relayMeter) loadData(from, to time.Time) error {
 			r.Logger.WithFields(logger.Fields{"error": err}).Warn("Error loading todays usage data")
 			return err
 		}
+
+		todaysLatency, err = r.Backend.TodaysLatency()
+		if err != nil {
+			r.Logger.WithFields(logger.Fields{"error": err}).Warn("Error loading todays latency data")
+			return err
```
Comment (adshmh) [inline api/meter.go]: I think an error in collecting latency data should not block other metrics from being retrieved.

File: api/meter.go [hunk 10]
```diff
@@ -163,18 +187,24 @@ func (r *relayMeter) loadData(from, to time.Time) error {
 
 	if updateDaily {
 		r.dailyUsage = dailyUsage
+
```
Comment (adshmh) [inline api/meter.go:210]: Please remove this extra blank line introduced by the PR.

File: api/meter.go [hunk 11]
```diff
@@ -163,18 +187,24 @@ func (r *relayMeter) loadData(from, to time.Time) error {
...
 	if updateToday {
 		r.todaysUsage = todaysUsage
+		r.todaysLatency = todaysLatency
+
 		d := r.RelayMeterOptions.TodaysMetricsTTL
 		if int(d.Seconds()) == 0 {
 			d = time.Duration(TTL_TODAYS_METRICS_DEFAULT_SECONDS) * time.Second
 		}
+
```
Comment (adshmh) [inline api/meter.go:228]: Same with this extra blank line, as well as previous ones in this function.

File: api/meter.go [hunk 12]
```diff
@@ -228,6 +258,38 @@ func (r *relayMeter) AppRelays(app string, from, to time.Time) (AppRelaysRespons
 	return resp, nil
 }
 
+func (r *relayMeter) AppLatency(app string) (AppLatencyResponse, error) {
+	r.Logger.WithFields(logger.Fields{"app": app}).Info("apiserver: Received AppLatency request")
+
+	resp := AppLatencyResponse{
```
Comment (adshmh) [inline api/meter.go]: Perhaps more readable to simply return this value instead of declaring a new variable that is used only once.

File: api/meter.go [hunk 13]
```diff
@@ -228,6 +258,38 @@ func (r *relayMeter) AppRelays(app string, from, to time.Time) (AppRelaysRespons
...
+	r.Logger.Info("apiserver: Received AllAppsLatency request")
+
+	resp := []AppLatencyResponse{}
+
+	for app, appLatency := range r.todaysLatency {
+		latencyResp := AppLatencyResponse{
+			Application:  app,
+			DailyLatency: appLatency,
+			From:         appLatency[0].Time,
+			To:           appLatency[23].Time,
```
Comment (adshmh) [inline api/meter.go]: Better to use the length of `appLatency` to avoid hard-coding the expected length.
Also are these latency entries guaranteed to be sorted based on Time?
Sorting the entries here is probably needed to avoid making any assumptions on returned data elsewhere.

File: api/meter.go [hunk 14]
```diff
@@ -567,16 +629,20 @@ func (r *relayMeter) StartDataLoader(ctx context.Context) {
...
 		r.Logger.WithFields(logger.Fields{"from": from, "to": to}).Info("Starting data loader...")
+
 		if err := r.loadData(from, to); err != nil {
 			r.Logger.WithFields(logger.Fields{"error": err}).Warn("Error setting timespan for data loader")
 		}
 	}
 
 	r.Logger.WithFields(logger.Fields{"maxArchiveAge": maxPastDays}).Info("Running initial data loader iteration...")
 	load(maxPastDays)
+
```
Comment (adshmh) [inline api/meter.go:756]: Are these blank lines added by the linter?

File: api/meter_test.go [hunk 15]
```diff
@@ -1074,6 +1076,11 @@ func (f *fakeBackend) TodaysUsage() (map[string]RelayCounts, error) {
 	return f.todaysUsage, nil
 }
 
+func (f *fakeBackend) TodaysLatency() (map[string][]Latency, error) {
+	f.todaysLatencyCalls++
+	return f.todaysLatency, nil
+}
```
Comment (adshmh) [inline api/meter_test.go]: Shouldn't the mock allow testing the error case? i.e. when the backend service returns an error?
This should also be covered by the unit tests.

File: api/server.go [hunk 16]
```diff
@@ -206,6 +222,16 @@ func GetHttpServer(meter RelayMeter, l *logger.Logger) func(w http.ResponseWrite
 			return
 		}
 
+		if appID := match(appsLatencyPath, req.URL.Path); appID != "" {
```
Comment (adshmh) [inline api/server.go]: These need to happen before the `totalRelays` match: otherwise they may not be reached.

File: cmd/apiserver/main.go [hunk 17]
```diff
@@ -12,91 +12,43 @@ import (
...
-	}
-
-	backendUrl := os.Getenv(ENV_BACKEND_API_URL)
-	if backendUrl == "" {
-		return options, fmt.Errorf("Missing required environment variable: %s", ENV_BACKEND_API_URL)
-	}
-	options.backendApiUrl = backendUrl
+var (
+	postgresUser     = environment.MustGetString("POSTGRES_USER")
+	postgresPassword = environment.MustGetString("POSTGRES_PASSWORD")
```
Comment (adshmh) [inline cmd/apiserver/main.go]: These changes do not belong in this PR. Could we remove them? I have submitted a review for the other PR.

**Review by adshmh — CHANGES_REQUESTED** [review]
Thank you for the PR!
- The approach looks good.
- Please remove all `tmp` directories and the corresponding binaries introduced by the PR.
- All changes unrelated to this PR (e.g. use of env. package, etc.) need to be moved to another PR (I think you already did this, i.e. PR #24, then simply removing the changes from this PR is all that is needed)

File: api/meter.go (hunk 11 above)
Comment (commoddity) [reply api/meter.go:228]: Added these ones for readability; personally I find it much easier to read functions when the logic is broken up a bit with some white space. Can remove if you disagree?

File: api/meter.go (hunk 10 above)
Comment (commoddity) [reply api/meter.go:210]: removed

File: api/meter.go (hunk 9 above)
Comment (commoddity) [reply api/meter.go]: No problem. Do you think a log is sufficient in this case? And should I remove the return only from the `TodaysLatency()` call or from `TodaysUsage()` as well?

File: api/meter.go (hunk 8 above)
Comment (commoddity) [reply api/meter.go]: Updated.

File: api/meter.go (hunk 9 above)
Comment (commoddity) [reply api/meter.go]: As the func existed already there was an error return for daily usage throwing an error which would have stopped collection of daily usage metrics. So just trying to understand if they should all be consistent or if only a failure of the latency should not stop the rest of the collection?

File: api/meter.go (hunk 12 above)
Comment (commoddity) [reply api/meter.go]: Updated

File: api/meter.go (hunk 13 above)
Comment (commoddity) [reply api/meter.go]: Updated to use the length of slice and ensure latency values are sorted by date in ascending order.

File: api/meter.go (hunk 14 above)
Comment (commoddity) [reply api/meter.go:756]: I added them to improve readability. I know this is a matter of opinion but I find logic much more readable when it's broken up logically with some whitespace. Not a sticking point for me though, can remove if you want.

File: api/meter_test.go (hunk 15 above)
Comment (commoddity) [reply api/meter_test.go]: Added the latency funcs to the unit tests, including testing backend errors.

File: api/server.go (hunk 16 above)
Comment (commoddity) [reply api/server.go]: Moved above the `totalRelays` match

File: cmd/apiserver/main.go (hunk 17 above)
Comment (commoddity) [reply cmd/apiserver/main.go]: While it's possible to remove these changes from the PR, the environment changes were necessary for me to be able to run relay meter on my workstation. 

This allowed me to test in a reproduction of the deployed application, giving me confidence that the changes worked in a more or less E2E environment.

Can we not simply merge https://github.com/pokt-foundation/relay-meter/pull/24 first, meaning the environment changes will not show up as a diff in this PR?

Comment (commoddity) [reply cmd/apiserver/main.go]: #24 is merged now so there's no longer any diff showing up on this PR between the `cmd/<service>/main.go` files and `main`

File: api/meter.go (hunk 9 above)
Comment (adshmh) [reply api/meter.go]: Good point: that is an inconsistency which needs fixing. In general errors in collecting any of the metrics should not have an effect on other metrics' collection.

File: api/meter.go (hunk 11 above)
Comment (adshmh) [reply api/meter.go:228]: I think this instance is fine to keep.

File: api/meter.go (hunk 14 above)
Comment (adshmh) [reply api/meter.go:756]: I think we can keep the extra blank lines if they are added intentionally to improve readability.

File: db/influx.go [hunk 18]
```diff
@@ -236,7 +230,8 @@ func (i *influxDB) TodaysLatency() (map[string][]api.Latency, error) {
 				}
 
 			}
-			latencyByHour := api.Latency{Time: hourlyTime, Latency: roundFloat(hourlyAverageLatency, 5)}
+
+			latencyByHour := api.Latency{Time: hourlyTime, Latency: numbers.RoundFloat(hourlyAverageLatency, 5)}
```
Comment (adshmh) [inline db/influx.go:299]: Please use GO's math library for rounding: https://pkg.go.dev/math

File: db/postgres.go [hunk 19]
```diff
@@ -386,7 +375,7 @@ func (p *pgClient) TodaysLatency() (map[string][]api.Latency, error) {
 			return nil, fmt.Errorf("Empty application public key, in query result line: %s", r)
 		}
 
-		latencyByHour := api.Latency{Time: hourlyTime, Latency: roundFloat(hourlyAverageLatency, 5)}
+		latencyByHour := api.Latency{Time: hourlyTime, Latency: numbers.RoundFloat(hourlyAverageLatency, 5)}
```
Comment (adshmh) [inline db/postgres.go:425]: Please use the standard math library from GO: https://pkg.go.dev/math

**Review by adshmh — CHANGES_REQUESTED** [review]
Thank you for updating the PR. Just a minor change. LGTM otherwise.

File: db/influx.go (hunk 18 above)
Comment (commoddity) [reply db/influx.go:299]: resolving per our convo on discord

File: db/postgres.go (hunk 19 above)
Comment (commoddity) [reply db/postgres.go:425]: resolving per our convo on discord

**Review by ricantar — DISMISSED** [review]
LGTM! thank you for solving the merge conflicts

File: api/meter.go [hunk 20]
```diff
@@ -250,6 +280,50 @@ func (r *relayMeter) AppRelays(app string, from, to time.Time) (AppRelaysRespons
...
+	for app, appLatency := range r.todaysLatency {
+		sort.Slice(appLatency, func(i, j int) bool {
+			return appLatency[i].Time.Before(appLatency[j].Time)
+		})
+
+		latencyResp := AppLatencyResponse{
+			Application:  app,
+			DailyLatency: appLatency,
+			From:         appLatency[0].Time,
```
Comment (adshmh) [inline api/meter.go]: Is it ever possible, e.g. due to a db error, to have an empty array as `appLatency`? If so, that could cause a panic here.

File: api/meter_test.go [hunk 21]
```diff
@@ -730,6 +731,148 @@ func TestAllAppsRelays(t *testing.T) {
...
+			}
+
+			relayMeter := NewRelayMeter(&fakeBackend, logger.New(), RelayMeterOptions{LoadInterval: 100 * time.Millisecond})
+			time.Sleep(200 * time.Millisecond)
+			got, err := relayMeter.AppLatency(tc.requestedApp)
+			if err != nil {
+				if tc.expectedErr == nil {
+					t.Fatalf("Unexpected error: %v", err)
+				}
+				if !strings.Contains(err.Error(), tc.expectedErr.Error()) {
```
Comment (adshmh) [inline api/meter_test.go:785]: Not a blocker, but checking for specific errors is preferable to looking inside the error string.

**Review by adshmh — CHANGES_REQUESTED** [review]
Thank you for the PR! Just one minor blocker.

File: api/meter.go (hunk 20 above)
Comment (commoddity) [reply api/meter.go]: Updated to avoid adding to response if Latency response slice is empty.

**Review by adshmh — APPROVED** [review]
LGTM. Thanks!
